infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/runtime/LogReplicationClientRouter.java

```diff
@@ -0,0 +1,432 @@
+                    }
+
+                    // Get Remote Leader
+                    if(runtimeFSM.getRemoteLeader().isPresent()) {
+                        endpoint = runtimeFSM.getRemoteLeader().get();
+                    } else {
+                        log.error("Leader not found to remote cluster {}", remoteClusterId);
+                        runtimeFSM.input(new LogReplicationRuntimeEvent(LogReplicationRuntimeEvent.LogReplicationRuntimeEventType.REMOTE_LEADER_LOSS));
+                        throw new ChannelAdapterException(
+                                String.format("Leader not found to remote cluster %s", remoteClusterDescriptor.getClusterId()));
+                    }
+                }
+
+                // In the case the message is intended for a specific endpoint, we do not
+                // block on connection future, this is the case of leader verification.
+                log.info("Send message to {}, type={}", endpoint, message.getMsgType());
+                channelAdapter.send(endpoint, CorfuMessageConverter.toProtoBuf(message));
+
+            } catch (Exception e) {
+                outstandingRequests.remove(requestId);
+                log.error("sendMessageAndGetCompletable: Remove request {} to {} due to exception! Message:{}",
+                        requestId, remoteClusterId, message, e);
+                cf.completeExceptionally(e);
+                return cf;
+            }
+
+            // Generate a timeout future, which will complete exceptionally
+            // if the main future is not completed.
+            final CompletableFuture<T> cfTimeout =
+                    CFUtils.within(cf, Duration.ofMillis(timeoutResponse));
+            cfTimeout.exceptionally(e -> {
+                if (e.getCause() instanceof TimeoutException) {
+                    outstandingRequests.remove(requestId);
+                    log.debug("sendMessageAndGetCompletable: Remove request {} to {} due to timeout! Message:{}",
+                            requestId, remoteClusterId, message);
+                }
+                return null;
+            });
+
+            return cfTimeout;
+        }
+
+        log.error("Invalid message type {}. Currently only log replication messages are processed.");
+        CompletableFuture<T> f = new CompletableFuture<>();
+        f.completeExceptionally(new Throwable("Invalid message type"));
+        return f;
+    }
+
+    /**
+     * Send a one way message, without adding a completable future.
+     *
+     * @param message The message to send.
+     */
+    @Override
+    public void sendMessage(CorfuMsg message) {
+        // Get the next request ID.
+        message.setRequestID(requestID.getAndIncrement());
+        // Get Remote Leader
+        if(runtimeFSM.getRemoteLeader().isPresent()) {
+            String remoteLeader = runtimeFSM.getRemoteLeader().get();
+            channelAdapter.send(remoteLeader, CorfuMessageConverter.toProtoBuf(message));
```

Comment:
If the connection is lost during channelAdapter.send(), should it detect it and generate an REMOTE_LEADER_LOSS message?

Reply:
No, the remote leader loss is a message to be sent from the standby not locally. If the connection is lost, we will get a connection_down from the leader and re-verify.